### PR TITLE
Fix __proto__ parsing in object literals

### PR DESCRIPTION
**Pull request by rerobika**
Co-authored-by: Rafal Walczyna r.walczyna@samsung.com
JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-parser-expr.c [hunk 1]
```diff
@@ -1029,9 +1029,10 @@ parser_parse_object_literal (parser_context_t *context_p) /**< context */
       default:
       {
 #if ENABLED (JERRY_ES2015)
-        if ((context_p->token.lit_location.type == LEXER_IDENT_LITERAL
-            || context_p->token.lit_location.type == LEXER_STRING_LITERAL)
-            && lexer_compare_literal_to_string (context_p, "__proto__", 9))
+        bool is_proto = ((context_p->token.lit_location.type == LEXER_IDENT_LITERAL
+                          || context_p->token.lit_location.type == LEXER_STRING_LITERAL)
+                         && lexer_compare_literal_to_string (context_p, "__proto__", 9));
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/parser/js/js-parser-expr.c]: Good catch, fixed!

File: jerry-core/parser/js/js-parser-expr.c [hunk 2]
```diff
@@ -1078,8 +1081,9 @@ parser_parse_object_literal (parser_context_t *context_p) /**< context */
           break;
         }
 
-        if (context_p->token.type == LEXER_RIGHT_BRACE
-            || context_p->token.type == LEXER_COMMA)
+        if ((context_p->token.type == LEXER_RIGHT_BRACE
+             || context_p->token.type == LEXER_COMMA)
+             && !is_proto)
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser-expr.c:1086]: Misaligned. Mayb e the two lines could be merged.

File: jerry-core/vm/vm.h [hunk 3]
```diff
@@ -271,6 +271,7 @@ typedef enum
   VM_OC_PUSH_NEW_TARGET,         /**< push new.target onto the stack */
   VM_OC_REQUIRE_OBJECT_COERCIBLE,/**< RequireObjectCoercible opretaion */
   VM_OC_ASSIGN_SUPER,            /**< assign super reference */
+  VM_OC_SET__PROTO__,            /**< set __proto__ */
```
Comment (zherczeg) [inline jerry-core/vm/vm.h:274]: Comment: set prototpe when `__proto__:` form is used

File: jerry-core/vm/vm.h [hunk 4]
```diff
@@ -333,6 +334,7 @@ typedef enum
   VM_OC_PUSH_NEW_TARGET = VM_OC_NONE,         /**< push new.target onto the stack */
   VM_OC_REQUIRE_OBJECT_COERCIBLE = VM_OC_NONE,/**< RequireObjectCoercible opretaion */
   VM_OC_ASSIGN_SUPER = VM_OC_NONE,            /**< assign super reference */
+  VM_OC_SET__PROTO__ = VM_OC_NONE,            /**< set __proto__ */
```
Comment (zherczeg) [inline jerry-core/vm/vm.h:337]: Ditto.

File: tests/jerry/es2015/object-prototype-proto.js [hunk 5]
```diff
@@ -54,5 +54,33 @@ try {
...
+var __proto__ = [];
+var o5 = { __proto__ };
+assert(o5.__proto__ === __proto__);
+assert(Object.getPrototypeOf(o5) === Object.prototype);
+
+var o6 = { __proto__() { return "42" } };
+assert(o6.__proto__() === "42");
+assert(Object.getPrototypeOf(o6) === Object.prototype);
+
+
```
Comment (zherczeg) [inline tests/jerry/es2015/object-prototype-proto.js:78]: One newline.

File: tests/jerry/es2015/object-prototype-proto.js [hunk 6]
```diff
@@ -54,5 +54,33 @@ try {
...
-var o2 =  { ["__proto__"] : null };
+var o2 = { ["__proto__"] : null };
 assert(o2.__proto__ === null);
+assert(Object.getPrototypeOf(o2) === Object.prototype);
+
+var o3 = { __proto__ : null };
+assert(o3.__proto__ === undefined);
+assert(Object.getPrototypeOf(o3) === null);
+
+var o4 = { "__proto__" : null };
```
Comment (zherczeg) [inline tests/jerry/es2015/object-prototype-proto.js:65]: There should be a test when the value is set to an existing object. Just create a `var o = {}`

**Comment by rerobika** [comment]
@zherczeg Thanks for the review, I applied your suggestions.

**Pull request merged**